File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        res = []
        count = 0
        
        for ele in s:
            if ele == "(":
                res.append(ele)
                count += 1
            elif ele == ")" and count > 0:
                res.append(ele)
                count -= 1
            elif ele != ")":
                res.append(ele)
                
        ans = []
        
        for i in res[::-1]:
            if i == "(" and count > 0:
                count -= 1
            else:
                ans.append(i)
                
        return ''.join(ans[::-1])

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        ArrayList<Character> arr = new ArrayList<>();
        char[] l = s.toCharArray();
        int count = 0;

        for (char ch : l) {
            if (ch == '(') {
                arr.add(ch);
                count++;
            } else if (ch == ')' && count > 0) {
                arr.add(ch);
                count--;
            } else if (ch != ')') {
                arr.add(ch);
            }
        }

        ArrayList<Character> res = new ArrayList<>();

        for (int i = arr.size()-1; i>-1 ; i--) {
            if (arr.get(i) == '(' && count > 0) {
                count--;
            } else {
                res.add(arr.get(i));
            }
        }

        StringBuilder ans = new StringBuilder();
        for (Character c : res) {
            ans.append(c);
        }

        return ans.reverse().toString();
    }
}